changehc/delphi_changehc/backfill.py

```diff
@@ -0,0 +1,119 @@
+    """
+    Store county level backfill data into backfill_dir.
+
+    Parameter:
+        df: pd.DataFrame
+            Store the county level raw file for backfill
+        _end_date: datetime
+            The most recent date when the raw data is received
+        backfill_dir: str
+            specified path to store backfill files.
+        numtype: str
+            indicate the type of the data
+        geo: str
+            geo level
+        weekday: bool
+    """
+    # We only need to run it once for a numtype
+    if geo != "county":
+        return
+    if weekday:
+        return
+
+    backfilldata = df.reset_index().copy()
+    backfilldata.rename({"timestamp": "time_value"}, axis=1, inplace=True)
+    #Store one year's backfill data
+    _start_date = _end_date.replace(year=_end_date.year-1)
+    selected_columns = ['time_value', 'fips',
+                        'num', 'den']
+    backfilldata = backfilldata.loc[backfilldata["time_value"] >= _start_date,
+                                    selected_columns]
+    path = backfill_dir + \
+        "/changehc_%s_as_of_%s.parquet"%(numtype, datetime.strftime(_end_date, "%Y%m%d"))
+    # Store intermediate file into the backfill folder
+    backfilldata.to_parquet(path)
+
+def merge_backfill_file(backfill_dir, numtype, geo, weekday, backfill_merge_day,
+                        today, test_mode=False, check_nd=25):
+    """
+    Merge ~4 weeks' backfill data into one file.
+
+    Usually this function should merge 28 days' data into a new file so as to
+    save the reading time when running the backfill pipelines. We
+    Parameters
+    ----------
+    today : datetime
+        The current dropdate
+    backfill_dir : str
+        specified path to store backfill files.
+    backfill_merge_day: int
+        The day of a week that we used to merge the backfill files. e.g. 0
+        is Monday.
+    test_mode: bool
+    check_nd: int
+        The criteria of the number of unmerged files. Ideally, we want the
+        number to be 28, but we use a looser criteria from practical
+        considerations
+    numtype: str
+            indicate the type of the data
+        geo: str
+            geo level
+        weekday: bool
+    """
+    # We only need to run it once for a numtype
+    if geo != "county":
+        return
+    if weekday:
+        return
+
+    new_files = glob.glob(backfill_dir + "/changehc_%s_as_of_*"%numtype)
+
+    def get_date(file_link):
+        # Keep the function here consistent with the backfill path in
+        # function `store_backfill_file`
+        fn = file_link.split("/")[-1].split(".parquet")[0].split("_")[-1]
+        return datetime.strptime(fn, "%Y%m%d")
+
+    date_list = list(map(get_date, new_files))
+    earliest_date = min(date_list)
+    latest_date = max(date_list)
+
+    # Check whether to merge
+    # Check the number of files that are not merged
+    if today.weekday() != backfill_merge_day or (today-earliest_date).days <= check_nd:
```

Comment:
You are right. Though the number of new files might not be the same as the number of days checked here, but either works.

Reply:
I didn't understand the second part that you mentioned. As for the backfill correction script, every time, need to read 300+ such daily backfill files for a single indicator, which take a long time to read. If we merge it to ~13-14 r-week combined files, it will save time to read the input data from my experience.